Replace autogenerated modelschema and enhance test design

Joshua-Douglas opened the pull request:
This PR replaces the autogenerated model schema classes with `marshmallow` schema objects. We are no longer using the OpenAPI code generation tool, and marshmallow provides a rich suite of serialization/deserialization tools that are capable of serializing complex database models.

In order to properly extract the autogenerated schema, I had to refactor the host controller, service provider controller, and service provider repository.

In the future we can simplify the use of marshmallow even further by using the [Flask-Marshmallow](https://flask-marshmallow.readthedocs.io/en/latest/) library in combination with [Flask-SQLAlchem](https://flask-sqlalchemy.palletsprojects.com/en/3.0.x/) library, but these libraries are not strictly necessary. 

While making these change I also extended the relevant test suites and improved the testing design. The following enhancements were made to the test suites:

1. Replace the disk-based temporary database with an in-memory only database, and ensure that this database is destroyed between tests
2. Remove the complex mocking infrastructure that was used for the housing repo test suite. We can conduct the tests more simply by using this in-memory temporary database
3. Simplify the simulated HTTP request code by taking advantage of the [WSGI test client](https://werkzeug.palletsprojects.com/en/2.3.x/test/#testing-wsgi-applications) features
4. Extend the endpoint tests to do more than check status codes. We are now also checking the returned values, and verifying that the database was altered as expected

paulespinosa commented on api/openapi_server/repositories/service_provider_repository.py:
```diff
@@ -20,121 +22,83 @@ def __init__(self, db_engine=None):
+                session.commit()
+                session.flush()
+                # primary keys are lazy loaded by default
+                _ = newProvider.id
```
Because I read [Autogenerated primary key attributes](https://docs.sqlalchemy.org/en/14/tutorial/orm_data_manipulation.html#autogenerated-primary-key-attributes), I thought the primary key is already available by default.

Also, the sentence above that section says changes are flushed out when `commmit` is called.

Joshua-Douglas replied:
primary key is already available, but only in the database. You need to refresh the python object before id field is populated. See my explanation above for more detail

A better way of refreshing was added in [1d8a319](https://github.com/hackforla/HomeUniteUs/pull/575/commits/1d8a319d4d549ac1424390e5c468d8689aba3f97)

paulespinosa commented on api/openapi_server/test/__init__.py:
```diff
@@ -2,81 +2,67 @@
         app = connexion.App(__name__)
         app.app.json_encoder = JSONEncoder
         app.add_api(get_bundled_specs(Path('openapi_server/openapi/openapi.yaml')),
                 arguments={'title': 'Home Unite Us'},
```
Unrelated to this PR: Turns out, this isn't used. The `arguments` argument is used to populate template variables in the openapi definition. However, there's no such template variable `title` in the openapi.yaml.

Joshua-Douglas replied:
Great catch. I removed it in [1d8a319](https://github.com/hackforla/HomeUniteUs/pull/575/commits/1d8a319d4d549ac1424390e5c468d8689aba3f97) since it doesn't do anything

paulespinosa reviewed (commented):
Thanks @Joshua-Douglas for working on this. 

The same treatment should be applied to the the `user_controller` and `auth_controller` too.

I'm having a hard time understanding where the boundaries of the architectural layers exist and where the models/schemas belong. @ju1es was starting to make these boundaries clearer by implementing a repository for the service provider. Is that something we can or should address in this PR?

Joshua-Douglas commented on api/openapi_server/controllers/host_controller.py:
```diff
@@ -1,40 +1,17 @@
+        new_host = Host(name = body["name"])
+        session.add(new_host)
+        session.commit()
+        _ = new_host.id
```
Thanks for calling this out! Short answer is yes, accessing the id does force `SQLAlchemy` to refresh the instance so that it is synchronized with the newly added/updated database row. You can verify this by commenting out the line and running `pytest` to see the failures.

This is a hacky way of forcing the refresh, so I took a dive into the `SQLAlchemy` docs to see if there is a better solution.

There are three things to be aware of here:

1. Returning the `SQLAlchemy` python instance like this detaches the instance from the session. This means the instance has no attachment to the database, unless it is attached to a new session. Any attempt to load an uninitialized property will result in a runtime exception, so you need to be careful to only detach fully refreshed instances.
2. Calling `commit` on a `SQLAlchemy` python instance will mark that instance as [expired](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#committing). This means that newly committed instances are out-of-date with the database, and they need to be refreshed in order to properly reflect all the properties (such as the id in our case). This refresh is normally done behind-the-scenes, but we need to force it since we are detaching it from the session.
3. Calls to `flush` will not update the SQLAlchemy python instance. Flushing transfers in-memory session changes to the database, without altering the SQLAlchemy python instance at all. 

So there are at least two options to ensure the id is available to the endpoint:

1. Do not detach the instance from the session. If the endpoint passes in a session to the repository method, then the repo method will still return an expired instance. This, however, won't be a problem because the expired instance will be attached to a session and therefore will be able to update the properties whenever they are needed. 
2. Instead of calling `_ = newProvider.id` to force the refresh of the id field, we can call `session.refresh(newProvider)` to guarantee that all properties are refreshed before detaching it from the instance. 

Option 1 would require passing a session instance into the repository methods. I think this would make using them much more complicated, so I'll use Option 2.

Joshua-Douglas replied:
Fix the refresh hack in [1d8a319](https://github.com/hackforla/HomeUniteUs/pull/575/commits/1d8a319d4d549ac1424390e5c468d8689aba3f97)

Joshua-Douglas commented on api/openapi_server/models/database.py:
```diff
@@ -1,10 +1,9 @@
 from sqlalchemy import Column, Integer, String, ForeignKey, DateTime, LargeBinary
-from sqlalchemy import text
-
+from marshmallow import Schema, fields
```
Great idea. Moved schema into their own file `models.schema` module in [1d8a319](https://github.com/hackforla/HomeUniteUs/pull/575/commits/1d8a319d4d549ac1424390e5c468d8689aba3f97)

Joshua-Douglas commented:
> Thanks @Joshua-Douglas for working on this.
> 
> The same treatment should be applied to the the `user_controller` and `auth_controller` too.
> 
> I'm having a hard time understanding where the boundaries of the architectural layers exist and where the models/schemas belong. @ju1es was starting to make these boundaries clearer by implementing a repository for the service provider. Is that something we can or should address in this PR?

Hey @paulespinosa, I think the PR is ready for re-review. I've addressed all your comments.

We discussed the confusion about the architectural layers and here is a brief summary:
- Models: SQLAlchemy database model objects represent tables stored in the database
- Model Schema: Marshmallow database model schema only exist to serialized / deserialize model objects as json. If you want to return a model object as json, then defining a schema makes the serialization easier
- OpenAPI Schema: The OpenAPI specification schema outline the expected request and response format. `connexion` uses this schema to validate the requests, and can also optionally validate the responses. Since we define these schema in our OpenAPI spec we can guarantee that our endpoints are only called if the request has all the required parameters (all nicely converted to the correct type too). 
   - We were previously autogenerating OpenAPI schema in code using a code generation tool, so that we could use the response schema. These schema were challenging to generate, read, and edit - so now we use Marshmallow. Marshmallow also have a number of other features (like SQLAlchemy integration) that we are not currently using, but may find useful in the future
-  Repositories: We can think of these as 'database helpers' that add a layer of abstraction to simplify common database updates. These repositories should help use avoid code duplication, and keep our endpoint logic simpler. They also have the added bonus of allowing us to easily test the SQLAlchemy logic used by our endpoints without making a HTTP request

I agree that `user_controller` and `auth_controller` could use a similar update, but I think there outside the scope of this PR since they don't use any model schema. Plus, I think I'll have to refactor `auth_controller` while I add endpoint authentication.

paulespinosa reviewed (commented):
Thank you for updating. Code is looking good.

For `user_controller.py` and `auth_controller.py`, this is a good time to make the use of "session" consistent. In this PR, the use went from`with Session(db_engine) as db_session:` to `with DataAccessLayer.session() as session:` Updating the remaining controllers to be consistent has the benefit of helping future devs know which style/pattern to follow.

Thank you for the detailed description of the parts involved. To add, since this API is currently a thin layer around the database, it is essentially a CRUD application. The data models from the client's perspective are essentially a 1:1 representation of the database models. This hung me up for a while because the architectural style I was looking for was to see layer specific models and clear separations between them. The "mental model" for the architectural layers, that I had in mind at the time, is illustrated in the [Respository Pattern](https://www.cosmicpython.com/book/chapter_02_repository.html) chapter in the Cosmic Python book. At the end there is a tip "If your app is just a simple CRUD (create-read-update-delete) wrapper around a database, then you don’t need a domain model or a repository." Which helps to validate the current design of the controllers without a repository. Makes sense.

Joshua-Douglas commented on api/requirements.txt:
Ah, I see. Yeah I thought `requirements-dev.txt` was the development only requirements. I also took a look at the `pip-compile` docs, and I also didn't see an option to reference the `requirements.txt` file within `requirements-dev.txt`.

Corrected in 5396674.

Joshua-Douglas commented on api/openapi_server/controllers/service_provider_controller.py:
```diff
@@ -1,93 +1,59 @@
+    database.
+    """
+    provider_name = body["provider_name"]
+    newProvider = housing_provider_repository.create_service_provider(provider_name)
```
Thanks for catching! I did a review of all the updated code and fixed all non snake_case variables in [a564391](https://github.com/hackforla/HomeUniteUs/pull/575/commits/a564391d8c5b1e4a8b4a307a50d57badaf0a9e43).

Joshua-Douglas commented on api/openapi_server/repositories/service_provider_repository.py:
```diff
@@ -20,121 +22,81 @@ def __init__(self, db_engine=None):
+        if the service provider already exists.
+        """
+        with DataAccessLayer.session() as session:
+            existingProvider = session.query(HousingProgramServiceProvider).filter(
```
Thanks for catching! I did a review of all the updated code and fixed all non snake_case variables in [a564391](https://github.com/hackforla/HomeUniteUs/pull/575/commits/a564391d8c5b1e4a8b4a307a50d57badaf0a9e43).

Joshua-Douglas commented:
Hey @paulespinosa, 

Good point on removing the `with Session(db_engine) as db_session:` pattern from `user_controller.py` and `auth_controller.py`. I made the update in `[bb09e0e](https://github.com/hackforla/HomeUniteUs/pull/575/commits/bb09e0e7f911c5c8b2888f79f7a544c5cd04ebbe)`. Now everyone is using the `DataAccessLayer` session, instead of accessing the db_engine directly. 

I believe all of your comments are addressed and this PR is ready for re-review.
